pkg/test/env/ports.go

```diff
@@ -85,63 +69,27 @@ type Ports struct {
 	PilotHTTPPort uint16
 }
 
-func allocPortBase(name uint16) uint16 {
-	base := portBase + name*portNum
-	for i := 0; i < 10; i++ {
-		if allPortFree(base, portNum) {
-			return base
-		}
-		base += maxTestNum * portNum
-	}
-	log.Println("could not find free ports, continue the test...")
-	return base
-}
-
-func allocEnvoyPortBase(name uint16) uint16 {
-	base := portBase + name*portNum
-	for i := 0; i < 10; i++ {
-		if allPortFree(base, envoyPortNum) {
-			return base
-		}
-		base += maxTestNum * portNum
-	}
-	log.Println("could not find free ports, continue the test...")
-	return base
-}
-
-func allPortFree(base uint16, ports uint16) bool {
-	for port := base; port < base+ports; port++ {
-		if IsPortUsed(port) {
-			log.Println("port is used ", port)
-			return false
-		}
-	}
-	return true
-}
-
 // NewPorts allocate all ports based on test id.
-func NewPorts(name uint16) *Ports {
-	base := allocPortBase(name)
+func NewPorts(t *testing.T) *Ports {
 	return &Ports{
-		ClientProxyPort: base,
-		ServerProxyPort: base + 1,
-		TCPProxyPort:    base + 2,
-		AdminPort:       base + 3,
-		ExtraPort:       base + 4,
-		BackendPort:     base + 5,
-		DiscoveryPort:   base + 6,
-		STSPort:         base + 7,
+		ClientProxyPort: portManager.ReservePortNumberOrFail(t),
+		ServerProxyPort: portManager.ReservePortNumberOrFail(t),
+		TCPProxyPort:    portManager.ReservePortNumberOrFail(t),
+		AdminPort:       portManager.ReservePortNumberOrFail(t),
+		ExtraPort:       portManager.ReservePortNumberOrFail(t),
+		BackendPort:     portManager.ReservePortNumberOrFail(t),
+		DiscoveryPort:   portManager.ReservePortNumberOrFail(t),
+		STSPort:         portManager.ReservePortNumberOrFail(t),
 	}
 }
 
 // NewEnvoyPorts allocate ports for Envoy
-func NewEnvoyPorts(ports *Ports, name uint16) *Ports {
-	base := allocEnvoyPortBase(name)
+func NewEnvoyPorts(t *testing.T, ports *Ports) *Ports {
 	return &Ports{
-		ClientProxyPort: base,
-		ServerProxyPort: base + 1,
-		TCPProxyPort:    base + 2,
-		AdminPort:       base + 3,
+		ClientProxyPort: portManager.ReservePortNumberOrFail(t),
```

Comment:
old method makes it very easy to run into port conflicts. New method reuse common code and removes port conflicts. This lets us do stress testing of these tests - I ran 256 concurrently